servant-io-streams/src/Servant/IO/Streams.hs

```diff
@@ -0,0 +1,31 @@
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -fno-cse #-}
+-- | This module exports 'ToSourceIO' and 'FromSourceIO' for 'IOStreams.InputStream'
+module Servant.IO.Streams where
+
+import           Control.Monad.IO.Class (liftIO)
+import           System.IO.Unsafe
+                 (unsafePerformIO)
+import qualified System.IO.Streams.Core as IOS
+import           Servant.API.Stream
+import qualified Servant.Types.SourceT       as S
+
+instance ToSourceIO a (IOS.InputStream a) where
+  toSourceIO src = S.SourceT ($ go)
+    where
+      go = S.Effect $ trans <$> IOS.read src
+
+      trans Nothing = S.Stop
+      trans (Just c) = S.Yield c go
+
+instance FromSourceIO a (IOS.InputStream a) where
+  {-# NOINLINE fromSourceIO #-}
+  fromSourceIO src = unsafePerformIO $ S.unSourceT src $ IOS.fromGenerator . gen
```

Comment:
Doesn't seem that outlandish for a streaming interface to require some initial stateful allocation... Any chance we could have variants of these typeclasses operating in a `MonadIO`?